File: titus/titus/lib1/model/reg.py
#!/usr/bin/env python

# Copyright (C) 2014  Open Data ("Open Data" refers to
# one or more of the following companies: Open Data Partners LLC,
# Open Data Research LLC, or Open Data Capital LLC.)
# 
# This file is part of Hadrian.
# 
# Licensed under the Hadrian Personal Use and Evaluation License (PUEL);
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://raw.githubusercontent.com/opendatagroup/hadrian/master/LICENSE
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import math

def np():
    import numpy
    return numpy

from titus.fcn import Fcn
from titus.fcn import LibFcn
from titus.signature import Sig
from titus.signature import Sigs
from titus.datatype import *
from titus.errors import *
from titus.util import callfcn, div
import titus.P as P
from titus.lib1.array import argLowestN
from titus.lib1.prob.dist import Chi2Distribution

provides = {}
def provide(fcn):
    provides[fcn.name] = fcn

prefix = "model.reg."

#################################################################### 

class Linear(LibFcn):
    name = prefix + "linear"
    sig  = Sigs([Sig([{"datum": P.Array(P.Double())}, {"model": P.WildRecord("M", {"coeff": P.Array(P.Double()), "const": P.Double()})}], P.Double()),
                 Sig([{"datum": P.Array(P.Double())}, {"model": P.WildRecord("M", {"coeff": P.Array(P.Array(P.Double())), "const": P.Array(P.Double())})}], P.Array(P.Double())),
                 Sig([{"datum": P.Map(P.Double())}, {"model": P.WildRecord("M", {"coeff": P.Map(P.Double()), "const": P.Double()})}], P.Double()),
                 Sig([{"datum": P.Map(P.Double())}, {"model": P.WildRecord("M", {"coeff": P.Map(P.Map(P.Double())), "const": P.Map(P.Double())})}], P.Map(P.Double()))])
    def __call__(self, state, scope, paramTypes, datum, model):
        coeffType = [x["type"] for x in paramTypes[1]["fields"] if x["name"] == "coeff"][0]

        if coeffType == {'items': 'double', 'type': 'array'}: #sig1
            coeff = model["coeff"] + [model["const"]]
            datum = np().array(datum + [1.0])
            if len(datum) != len(coeff):
                raise PFARuntimeException("misaligned coeff")
            return float(np().dot(coeff, datum))

        elif coeffType == {'items': {'items': 'double', 'type': 'array'}, 'type': 'array'}: #sig2
            coeff = np().array(model["coeff"])
            const = np().array(model["const"])
            datum = np().array(datum + [1.0])
            try:
                coeff = np().vstack((coeff.T, const))
            except:
                raise PFARuntimeException('misaligned coeff')
            return map(float, np().dot(coeff.T, datum))

        elif coeffType == {'values': 'double', 'type': 'map'}: #sig3
            coeff = model["coeff"]
            const = model["const"]
            if len(datum) != len(coeff):
                raise PFARuntimeException("misaligned coeff")
            out = 0.0
            try:
                for key in datum.keys():
                    out += datum[key] * coeff[key]
            except:
                raise PFARuntimeException("misaligned const")
            return float(out + const)

        else:
            coeff = model["coeff"]
            const = model["const"]
            outMap = {}
            for ckey in coeff.keys():
                out = 0.0
                for rkey in coeff[ckey].keys():
                    try:
                        out += datum[rkey] * coeff[ckey][rkey]
                    except:
                        raise PFARuntimeException("misaligned coeff")
                try:
                    outMap[ckey] = float(out + const[ckey])
                except:
                    raise PFARuntimeException("misaligned const")
            return outMap
provide(Linear())

class LinearVariance(LibFcn):
    name = prefix + "linearVariance"
    sig = Sigs([Sig([{"datum": P.Array(P.Double())}, {"model": P.WildRecord("M", {"covar": P.Array(P.Array(P.Double()))})}], P.Double()),
                Sig([{"datum": P.Array(P.Double())}, {"model": P.WildRecord("M", {"covar": P.Array(P.Array(P.Array(P.Double())))})}], P.Array(P.Double())),
                Sig([{"datum": P.Map(P.Double())}, {"model": P.WildRecord("M", {"covar": P.Map(P.Map(P.Double()))})}], P.Double()),
                Sig([{"datum": P.Map(P.Double())}, {"model": P.WildRecord("M", {"covar": P.Map(P.Map(P.Map(P.Double())))})}], P.Map(P.Double()))])
    def __call__(self, state, scope, paramTypes, datum, model):
        covarType = [x["type"] for x in paramTypes[1]["fields"] if x["name"] == "covar"][0]

        if covarType == {"type": "array", "items": {"type": "array", "items": "double"}}:  # sig1
            datum = datum + [1.0]
            covar = model["covar"]
            if len(datum) != len(covar) or any(len(datum) != len(x) for x in covar):
                raise PFARuntimeException("misaligned covariance")
            x = np().matrix([datum])
            C = np().matrix(covar)
            return float(x.dot(C.dot(x.T))[0][0])

        elif covarType == {"type": "array", "items": {"type": "array", "items": {"type": "array", "items": "double"}}}:  # sig2
            datum = datum + [1.0]
            x = np().matrix([datum])
            out = []
            for covar in model["covar"]:
                if len(datum) != len(covar) or any(len(datum) != len(x) for x in covar):
                    raise PFARuntimeException("misaligned covariance")
                C = np().matrix(covar)
                out.append(float(x.dot(C.dot(x.T))[0][0]))
            return out

        elif covarType == {"type": "map", "values": {"type": "map", "values": "double"}}:  # sig3
            datum = dict(list(datum.items()) + [("", 1.0)])
            covar = model["covar"]
            keys = list(datum.keys())
            x = np().matrix([[datum[k] for k in keys]])
            try:
                C = np().matrix([[covar[i][j] for j in keys] for i in keys])
            except KeyError:
                raise PFARuntimeException("misaligned covariance")
            return float(x.dot(C.dot(x.T))[0][0])

        else:
            datum = dict(list(datum.items()) + [("", 1.0)])
            keys = list(datum.keys())
            x = np().matrix([[datum[k] for k in keys]])
            out = {}
            for depkey, covar in model["covar"].items():
                try:
                    C = np().matrix([[covar[i][j] for j in keys] for i in keys])
                except KeyError:
                    raise PFARuntimeException("misaligned covariance")
                out[depkey] = float(x.dot(C.dot(x.T))[0][0])
            return out

provide(LinearVariance())


class Residual(LibFcn):
    name = prefix + "residual"
    sig = Sigs([Sig([{"observation": P.Double()}, {"prediction": P.Double()}], P.Double()),
                Sig([{"observation": P.Array(P.Double())}, {"prediction": P.Array(P.Double())}], P.Array(P.Double())),
                Sig([{"observation": P.Map(P.Double())}, {"prediction": P.Map(P.Double())}], P.Map(P.Double()))])
    def __call__(self, state, scope, paramTypes, observation, prediction):
        if isinstance(observation, dict):
            if len(observation) != len(prediction):
                raise PFARuntimeException("misaligned prediction")
            result = {}
            for k, o in observation.items():
                try:
                    result[k] = o - prediction[k]
                except KeyError:
                    raise PFARuntimeException("misaligned prediction")
            return result

        elif isinstance(observation, (tuple, list)):
            try:
                result = [float(o - p) for o, p in zip(observation, prediction)]
            except:
                raise PFARuntimeException("misaligned prediction")
            return result

        else:
            return float(observation - prediction)

provide(Residual())

class Pull(LibFcn):
    name = prefix + "pull"
    sig = Sigs([Sig([{"observation": P.Double()}, {"prediction": P.Double()}, {"uncertainty": P.Double()}], P.Double()),
                Sig([{"observation": P.Array(P.Double())}, {"prediction": P.Array(P.Double())}, {"uncertainty": P.Array(P.Double())}], P.Array(P.Double())),
                Sig([{"observation": P.Map(P.Double())}, {"prediction": P.Map(P.Double())}, {"uncertainty": P.Map(P.Double())}], P.Map(P.Double()))])
    def __call__(self, state, scope, paramTypes, observation, prediction, uncertainty):
        if isinstance(observation, dict):
            if len(observation) != len(prediction):
                raise PFARuntimeException("misaligned prediction")
            if len(observation) != len(uncertainty):
                raise PFARuntimeException("misaligned uncertainty")
            result = {}
            for k, o in observation.items():
                try:
                    p = prediction[k]
                except KeyError:
                    raise PFARuntimeException("misaligned prediction")
                try:
                    u = uncertainty[k]
                except KeyError:
                    raise PFARuntimeException("misaligned uncertainty")
                try:
                    result[k] = (o - p) / u
                except ZeroDivisionError:
                    result[k] = float("nan")
            return result

        elif isinstance(observation, (tuple, list)):
            if len(observation) != len(prediction):
                raise PFARuntimeException("misaligned prediction")
            if len(observation) != len(uncertainty):
                raise PFARuntimeException("misaligned prediction")
            return [float((o - p)/u) for o, p, u in zip(observation, prediction, uncertainty)]

        else:
            return float((observation - prediction)/uncertainty)

provide(Pull())

class Mahalanobis(LibFcn):
    name = prefix + "mahalanobis"
    sig = Sigs([Sig([{"observation": P.Array(P.Double())}, {"prediction": P.Array(P.Double())},
                     {"covariance": P.Array(P.Array(P.Double()))}], P.Double()),
                Sig([{"observation": P.Map(P.Double())}, {"prediction": P.Map(P.Double())},
                     {"covariance": P.Map(P.Map(P.Double()))}], P.Double())])
    def __call__(self, state, scope, paramTypes, observation, prediction, covariance):
        if isinstance(observation, (tuple, list)):
            if (len(observation) < 1):
                raise PFARuntimeException("too few rows/cols")
            if (len(observation) != len(prediction)):
                raise PFARuntimeException("misaligned prediction")
            if (not all(len(i)==len(covariance[0]) for i in covariance)) and (len(covariance) != len(covariance[0])):
                raise PFARuntimeException("misaligned covariance")
            x = np().array([(o - p) for o, p in zip(observation, prediction)])
            C = np().array(covariance)
        else:
            if (len(observation) < 1):
                raise PFARuntimeException("too few rows/cols")
            if (len(observation) != len(prediction)):
                raise PFARuntimeException("misaligned prediction")
            # use observation keys throughout
            keys = observation.keys()
            try:
                x = np().array([observation[key] - prediction[key] for key in keys])
            except:
                raise PFARuntimeException("misaligned prediction")
            C = np().empty((len(keys), len(keys)))
            try:
                for i,k1 in enumerate(keys):
                    for j,k2 in enumerate(keys):
                        C[i,j] = float(covariance[k1][k2])
            except:
                raise PFARuntimeException("misaligned covariance")
        return float(np().sqrt(np().linalg.solve(C, x).T.dot(x)))
provide(Mahalanobis())


class UpdateChi2(LibFcn):
    name = prefix + "updateChi2"
    sig = Sigs([Sig([{"pull": P.Double()}, {"state_": P.WildRecord("A", {"chi2": P.Double(), "DOF": P.Int()})}], P.Wildcard("A")),
                Sig([{"pull": P.Array(P.Double())}, {"state_": P.WildRecord("A", {"chi2": P.Double(), "DOF": P.Int()})}], P.Wildcard("A")),
                Sig([{"pull": P.Map(P.Double())}, {"state_": P.WildRecord("A", {"chi2": P.Double(), "DOF": P.Int()})}], P.Wildcard("A"))])
    def __call__(self, state, scope, paramTypes, pull, state_):
        if isinstance(pull, float):
            return update(pull*pull, state_)
        elif isinstance(pull, (tuple, list)):
            return update(sum([y**2 for y in pull]), state_)
        else:
            return update(sum([y**2 for y in pull.values()]), state_)
provide(UpdateChi2())
def update(x, state_):
    state_["chi2"] = float(state_["chi2"] + x)
    state_["DOF"]  = float(state_["DOF"] + 1)
    return state_

class ReducedChi2(LibFcn):
    name = prefix + "reducedChi2"
    sig = Sig([{"state_": P.WildRecord("A", {"chi2": P.Double(), "DOF": P.Int()})}], P.Double())
    def __call__(self, state, scope, paramTypes, state_):
        return float(state_["chi2"]/state_["DOF"])
provide(ReducedChi2())

class Chi2Prob(LibFcn):
    name = prefix + "chi2Prob"
    sig = Sig([{"state_": P.WildRecord("A", {"chi2": P.Double(), "DOF": P.Int()})}], P.Double())
    def __call__(self, state, scope, paramTypes, state_):
        return float(Chi2Distribution(state_["DOF"]).CDF(state_["chi2"]))
provide(Chi2Prob())

File: hadrian/src/test/scala/lib1/model/reg.scala
// Copyright (C) 2014  Open Data ("Open Data" refers to
// one or more of the following companies: Open Data Partners LLC,
// Open Data Research LLC, or Open Data Capital LLC.)
// 
// This file is part of Hadrian.
// 
// Licensed under the Hadrian Personal Use and Evaluation License (PUEL);
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://raw.githubusercontent.com/opendatagroup/hadrian/master/LICENSE
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package test.scala.lib1.model.reg

import scala.collection.JavaConversions._
import scala.language.postfixOps

import org.junit.runner.RunWith

import org.scalatest.FlatSpec
import org.scalatest.junit.JUnitRunner
import org.scalatest.Matchers

import com.opendatagroup.hadrian.data._
import com.opendatagroup.hadrian.jvmcompiler._
import com.opendatagroup.hadrian.errors._
import test.scala._

@RunWith(classOf[JUnitRunner])
class Lib1ModelRegSuite extends FlatSpec with Matchers {
  "model.reg.linear" must "do one-level array signature" taggedAs(Lib1, Lib1ModelReg) in {
    val engine = PFAEngine.fromYaml("""
input: {type: array, items: double}
output: double
cells:
  model:
    type:
      type: record
      name: Model
      fields:
        - {name: coeff, type: {type: array, items: double}}
        - {name: const, type: double}
    init:
      coeff: [1, 2, 3, 0, 5]
      const: 100.0
action:
  model.reg.linear:
    - input
    - cell: model
""").head
    engine.action(engine.jsonInput("""[0.1, 0.2, 0.3, 0.4, 0.5]""")).asInstanceOf[java.lang.Double].doubleValue should be (103.9 +- 0.1)
  }

  it must "do two-level array signature" taggedAs(Lib1, Lib1ModelReg) in {
    val engine = PFAEngine.fromYaml("""
input: {type: array, items: double}
output: {type: array, items: double}
cells:
  model:
    type:
      type: record
      name: Model
      fields:
        - {name: coeff, type: {type: array, items: {type: array, items: double}}}
        - {name: const, type: {type: array, items: double}}
    init:
      coeff: [[1, 2, 3, 0, 5],
              [1, 1, 1, 1, 1],
              [0, 0, 0, 0, 1]]
      const: [0.0, 0.0, 100.0]
action:
  model.reg.linear:
    - input
    - cell: model
""").head
    val out = engine.action(engine.jsonInput("""[0.1, 0.2, 0.3, 0.4, 0.5]""")).asInstanceOf[PFAArray[Double]].toVector
    out(0) should be (3.9 +- 0.1)
    out(1) should be (1.5 +- 0.1)
    out(2) should be (100.5 +- 0.1)
  }

  it must "do one-level map signature" taggedAs(Lib1, Lib1ModelReg) in {
    val engine = PFAEngine.fromYaml("""
input: {type: map, values: double}
output: double
cells:
  model:
    type:
      type: record
      name: Model
      fields:
        - {name: coeff, type: {type: map, values: double}}
        - {name: const, type: double}
    init:
      coeff: {one: 1, two: 2, three: 3, four: 0, five: 5}
      const: 100.0
action:
  model.reg.linear:
    - input
    - cell: model
""").head
    engine.action(engine.jsonInput("""{"one": 0.1, "two": 0.2, "three": 0.3, "four": 0.4, "five": 0.5}""")).asInstanceOf[java.lang.Double].doubleValue should be (103.9 +- 0.1)
  }

  it must "do two-level map signature" taggedAs(Lib1, Lib1ModelReg) in {
    val engine = PFAEngine.fromYaml("""
input: {type: map, values: double}
output: {type: map, values: double}
cells:
  model:
    type:
      type: record
      name: Model
      fields:
        - {name: coeff, type: {type: map, values: {type: map, values: double}}}
        - {name: const, type: {type: map, values: double}}
    init:
      coeff:
        uno: {one: 1, two: 2, three: 3, four: 0, five: 5}
        dos: {one: 1, two: 1, three: 1, four: 1, five: 1}
        tres: {one: 0, two: 0, three: 0, four: 0, five: 1}
      const:
        {uno: 0.0, dos: 0.0, tres: 100.0}
action:
  model.reg.linear:
    - input
    - cell: model
""").head
    val out = engine.action(engine.jsonInput("""{"one": 0.1, "two": 0.2, "three": 0.3, "four": 0.4, "five": 0.5}""")).asInstanceOf[PFAMap[java.lang.Double]].toMap
    out("uno").doubleValue should be (3.9 +- 0.1)
    out("dos").doubleValue should be (1.5 +- 0.1)
    out("tres").doubleValue should be (100.5 +- 0.1)
  }

  "model.reg.linearVariance" must "do one-level array signature" taggedAs(Lib1, Lib1ModelReg) in {
    val engine = PFAEngine.fromYaml("""
input: {type: array, items: double}
output: double
cells:
  model:
    type:
      type: record
      name: Model
      fields:
        - {name: covar, type: {type: array, items: {type: array, items: double}}}
    init:
      covar: [[ 1.0, -0.1, 0.0],
              [-0.1,  2.0, 0.0],
              [ 0.0,  0.0, 0.0]]
action:
  model.reg.linearVariance:
    - input
    - cell: model
""").head
    engine.action(engine.jsonInput("""[0.1, 0.2]""")).asInstanceOf[java.lang.Double].doubleValue should be (0.086 +- 0.001)
  }

  it must "do two-level array signature" taggedAs(Lib1, Lib1ModelReg) in {
    val engine = PFAEngine.fromYaml("""
input: {type: array, items: double}
output: {type: array, items: double}
cells:
  model:
    type:
      type: record
      name: Model
      fields:
        - {name: covar, type: {type: array, items: {type: array, items: {type: array, items: double}}}}
    init:
      covar:
        - [[ 1.0, -0.1, 0.0],
           [-0.1,  2.0, 0.0],
           [ 0.0,  0.0, 0.0]]
        - [[ 1.0, -0.1, 0.0],
           [-0.1,  2.0, 0.0],
           [ 0.0,  0.0, 0.0]]
        - [[ 1.0, -0.1, 0.0],
           [-0.1,  2.0, 0.0],
           [ 0.0,  0.0, 0.0]]
        - [[ 1.0, -0.1, 0.0],
           [-0.1,  2.0, 0.0],
           [ 0.0,  0.0, 0.0]]
action:
  model.reg.linearVariance:
    - input
    - cell: model
""").head
    val results = engine.action(engine.jsonInput("""[0.1, 0.2]""")).asInstanceOf[PFAArray[Double]].toVector
    results(0) should be (0.086 +- 0.001)
    results(1) should be (0.086 +- 0.001)
    results(2) should be (0.086 +- 0.001)
    results(3) should be (0.086 +- 0.001)
  }

  it must "do one-level map signature" taggedAs(Lib1, Lib1ModelReg) in {
    val engine = PFAEngine.fromYaml("""
input: {type: map, values: double}
output: double
cells:
  model:
    type:
      type: record
      name: Model
      fields:
        - {name: covar, type: {type: map, values: {type: map, values: double}}}
    init:
      covar: {a: {a:  1.0, b: -0.1, "": 0.0},
              b: {a: -0.1, b:  2.0, "": 0.0},
             "": {a:  0.0, b:  0.0, "": 0.0}}
action:
  model.reg.linearVariance:
    - input
    - cell: model
""").head
    engine.action(engine.jsonInput("""{"a": 0.1, "b": 0.2}""")).asInstanceOf[java.lang.Double].doubleValue should be (0.086 +- 0.001)
  }

  it must "do two-level map signature" taggedAs(Lib1, Lib1ModelReg) in {
    val engine = PFAEngine.fromYaml("""
input: {type: map, values: double}
output: {type: map, values: double}
cells:
  model:
    type:
      type: record
      name: Model
      fields:
        - {name: covar, type: {type: map, values: {type: map, values: {type: map, values: double}}}}
    init:
      covar:
        one: {a: {a:  1.0, b: -0.1, "": 0.0},
              b: {a: -0.1, b:  2.0, "": 0.0},
             "": {a:  0.0, b:  0.0, "": 0.0}}
        two: {a: {a:  1.0, b: -0.1, "": 0.0},
              b: {a: -0.1, b:  2.0, "": 0.0},
             "": {a:  0.0, b:  0.0, "": 0.0}}
        three: {a: {a:  1.0, b: -0.1, "": 0.0},
                b: {a: -0.1, b:  2.0, "": 0.0},
               "": {a:  0.0, b:  0.0, "": 0.0}}
        four: {a: {a:  1.0, b: -0.1, "": 0.0},
               b: {a: -0.1, b:  2.0, "": 0.0},
              "": {a:  0.0, b:  0.0, "": 0.0}}
action:
  model.reg.linearVariance:
    - input
    - cell: model
""").head
    val results = engine.action(engine.jsonInput("""{"a": 0.1, "b": 0.2}""")).asInstanceOf[PFAMap[java.lang.Double]].toMap
    results("one").doubleValue should be (0.086 +- 0.001)
    results("two").doubleValue should be (0.086 +- 0.001)
    results("three").doubleValue should be (0.086 +- 0.001)
    results("four").doubleValue should be (0.086 +- 0.001)
  }

  "model.reg.residual" must "work" taggedAs(Lib1, Lib1ModelReg) in {
    val engine1 = PFAEngine.fromYaml("""
input: double
output: double
action:
  model.reg.residual:
    - input
    - 3.0
""").head
    engine1.action(java.lang.Double.valueOf(5.0)).asInstanceOf[java.lang.Double].doubleValue should be (2.0 +- 0.1)

    val engine2 = PFAEngine.fromYaml("""
input: {type: array, items: double}
output: {type: array, items: double}
action:
  model.reg.residual:
    - input
    - type: {type: array, items: double}
      value: [2.5, 2.5, 2.5]
""").head
    val out2 = engine2.action(engine2.jsonInput("""[1, 2, 3]""")).asInstanceOf[PFAArray[Double]].toVector
    out2(0) should be (-1.5 +- 0.1)
    out2(1) should be (-0.5 +- 0.1)
    out2(2) should be (0.5 +- 0.1)

    val engine3 = PFAEngine.fromYaml("""
input: {type: map, values: double}
output: {type: map, values: double}
action:
  model.reg.residual:
    - input
    - type: {type: map, values: double}
      value: {one: 2.5, two: 2.5, three: 2.5}
""").head
    val out3 = engine3.action(engine3.jsonInput("""{"one": 1, "two": 2, "three": 3}""")).asInstanceOf[PFAMap[java.lang.Double]].toMap
    out3("one").doubleValue should be (-1.5 +- 0.1)
    out3("two").doubleValue should be (-0.5 +- 0.1)
    out3("three").doubleValue should be (0.5 +- 0.1)
  }

  "model.reg.pull" must "work" taggedAs(Lib1, Lib1ModelReg) in {
    val engine1 = PFAEngine.fromYaml("""
input: double
output: double
action:
  model.reg.pull:
    - input
    - 3.0
    - 2.0
""").head
    engine1.action(java.lang.Double.valueOf(5.0)).asInstanceOf[java.lang.Double].doubleValue should be (1.0 +- 0.1)

    val engine2 = PFAEngine.fromYaml("""
input: {type: array, items: double}
output: {type: array, items: double}
action:
  model.reg.pull:
    - input
    - type: {type: array, items: double}
      value: [2.5, 2.5, 2.5]
    - type: {type: array, items: double}
      value: [2.0, 2.0, 2.0]
""").head
    val out2 = engine2.action(engine2.jsonInput("""[1, 2, 3]""")).asInstanceOf[PFAArray[Double]].toVector
    out2(0) should be (-0.75 +- 0.1)
    out2(1) should be (-0.25 +- 0.1)
    out2(2) should be (0.25 +- 0.1)

    val engine3 = PFAEngine.fromYaml("""
input: {type: map, values: double}
output: {type: map, values: double}
action:
  model.reg.pull:
    - input
    - type: {type: map, values: double}
      value: {one: 2.5, two: 2.5, three: 2.5}
    - type: {type: map, values: double}
      value: {one: 2.0, two: 2.0, three: 2.0}
""").head
    val out3 = engine3.action(engine3.jsonInput("""{"one": 1, "two": 2, "three": 3}""")).asInstanceOf[PFAMap[java.lang.Double]].toMap
    out3("one").doubleValue should be (-0.75 +- 0.1)
    out3("two").doubleValue should be (-0.25 +- 0.1)
    out3("three").doubleValue should be (0.25 +- 0.1)
  }

  "model.reg.mahalanobis" must "work" taggedAs(Lib1, Lib1ModelReg) in {
    val engine2 = PFAEngine.fromYaml("""
input: {type: array, items: double}
output: double
action:
  model.reg.mahalanobis:
    - input
    - type: {type: array, items: double}
      value: [2.5, 2.5, 2.5]
    - type: {type: array, items: {type: array, items: double}}
      value: [[2.0, 0.0, 0.0],
              [0.0, 4.0, 0.0],
              [0.0, 0.0, 1.0]]
""").head
    engine2.action(engine2.jsonInput("""[1, 2, 3]""")).asInstanceOf[java.lang.Double].doubleValue should be (1.19895788083 +- 0.1)

    val engine3 = PFAEngine.fromYaml("""
input: {type: map, values: double}
output: double
action:
  model.reg.mahalanobis:
    - input
    - type: {type: map, values: double}
      value: {one: 2.5, two: 2.5, three: 2.5}
    - type: {type: map, values: {type: map, values: double}}
      value: {one:   {one: 2.0, two: 0.0, three: 0.0},
              two:   {one: 0.0, two: 4.0, three: 0.0},
              three: {one: 0.0, two: 0.0, three: 1.0}}
""").head
    engine3.action(engine3.jsonInput("""{"one": 1, "two": 2, "three": 3}""")).asInstanceOf[java.lang.Double].doubleValue should be (1.19895788083 +- 0.1)
  }

  "model.reg.*chi2*" must "work" taggedAs(Lib1, Lib1ModelReg) in {
    val engine = PFAEngine.fromYaml("""
input: "null"
output: {type: array, items: double}
action:
  - let:
      state:
        type:
          type: record
          name: Chi2
          fields:
            - {name: chi2, type: double}
            - {name: DOF, type: int}
        new:
          {chi2: 0.0, DOF: -4}
  - set: {state: {model.reg.updateChi2: [0.0, state]}}
  - set: {state: {model.reg.updateChi2: [1.0, state]}}
  - set: {state: {model.reg.updateChi2: [1.5, state]}}
  - set: {state: {model.reg.updateChi2: [-0.75, state]}}
  - set: {state: {model.reg.updateChi2: [-1.0, state]}}
  - set: {state: {model.reg.updateChi2: [0.5, state]}}
  - set: {state: {model.reg.updateChi2: [-1.5, state]}}
  - type: {type: array, items: double}
    new: [state.chi2, {model.reg.reducedChi2: state}, {model.reg.chi2Prob: state}]
""").head
    val out = engine.action(null).asInstanceOf[PFAArray[Double]].toVector
    out(0) should be (7.3125 +- 0.1)
    out(1) should be (2.4375 +- 0.1)
    out(2) should be (0.9374 +- 0.1)
  }
}
